test/core/test_poll.c

```diff
@@ -9,14 +9,35 @@
 #include <errno.h>
 #include <fcntl.h>
 #include <poll.h>
+#include <emscripten.h>
 
 int main() {
+  EM_ASM(
+    var dummy_device = FS.makedev(64, 0);
+#if WASMFS
+    // WasmFS has different requirements for a device's functions compared
+    // to the legacy API.
```

Comment:
Can we reduce these differences? Ideally they would not exist.

Reply:
I was thinking about that, but the old FS API allows you to pass custom methods like `llseek` or `open` to devices, whereas those methods aren't accepted/used by the JS IMPL backend (e.g. in `test_64bit.c`). Another difference is that the old FS API allows you to pass an empty set of functions to a device in order to use default behavior (Line 28 of `test_poll.c`), but I think JS IMPL doesn't really have the concept of "default behavior", so we have to at least pass in empty methods for `allocFile`, `freeFile`, etc. 

To keep full compatibility I considered "overriding" the JS methods we provide in `library_wasmfs.js` when a user passes in definitions for device functions like `open` or `llseek`, but this would probably mean the user's custom functions wouldn't interact with WasmFS, which might cause some issues.

Reply:
Ah, this makes sense. It looks like the old FS threw `EINVAL` errors when it attempts to call methods that aren't provided. My current approach was to add empty method bodies for `allocFile`, `freeFile`, and `getSize` by default (Since they didn't exist in the old FS) and throw `EINVAL` when the user attempts to call `read` or `write` without passing in the correct values.

However, I'm a little stuck making the WasmFS API `read` and `write` functions match the old API. The original `read` and `write` methods received an entire array object for `buffer` with an offset to start writing at, but the JSIMPL methods take in a pointer to the backend. I tried logging the values and it looks like the old API's `offset` values and the WasmFS API's `buffer` pointer point to different locations. Should they be pointing at the same location in the shared memory between JS and Wasm?

Reply:
Ah, they might not point to the same place. A JS typed array view has its own internal offset. So if the absolute address of a file in memory was 1000, but we took a view starting from 800, then the offset relative to that view would be 200.

I still hope we can translate the old form to the new? You can do `.byteOffset` on a typed array view to get its offset in the entire space, which might help here.